web/packages/teleterm/src/mainProcess/profileWatcher/profileWatcher.test.ts

```diff
@@ -0,0 +1,347 @@
+  // Insert two rapid events within debounce interval.
+  await tshClientMock.insertOrUpdateCluster(cluster);
+  await tshClientMock.insertOrUpdateCluster(cluster);
+  // Wait slightly longer than debounce interval to ensure a single handler is called.
+  await wait(testDebounceMs + testDebounceMs / 2);
+  expect(handler).toHaveBeenCalledTimes(1);
+  handler.mockClear();
+
+  // Test no events are lost when processing is slow.
+  handler.mockImplementation(() => wait(2 * testDebounceMs));
+  await tshClientMock.insertOrUpdateCluster(cluster);
+  // Insert the second event while the first event is still processed
+  // (it will finish at 2*testDebounceMs).
+  await wait(testDebounceMs + testDebounceMs / 2);
+  await tshClientMock.insertOrUpdateCluster(cluster);
+  await expect(() => {
+    return handler.mock.calls.length === 2;
+  }).toEventuallyBeTrue({ waitFor: 1000, tick: 10 });
+});
+
+test('watcher stops when consumer throws', async () => {
+  const tshDir = await makePerTestDir();
+  const tshClientMock = await mockTshClient(tshDir, { clusters: [] });
+  const clusterStoreMock = mockClusterStore({ clusters: [] });
+  const watcher = watchProfiles({
+    tshDirectory: tshDir,
+    tshClient: tshClientMock,
+    clusterStore: clusterStoreMock,
+    signal: abortController.signal,
+    debounceMs: testDebounceMs,
+  });
+
+  await expect(watcher.throw(new Error('Consumer failure'))).rejects.toThrow(
+    'Consumer failure'
+  );
+
+  await tshClientMock.insertOrUpdateCluster(makeRootCluster());
+
+  const race = await Promise.race([
+    watcher.next(),
+    wait(300).then(() => 'timeout'),
+  ]);
+
+  expect(race).toStrictEqual({ done: true, value: undefined });
+});
+
+test('removing tsh directory does not break watcher', async () => {
+  const tshDir = await makePerTestDir();
+  const cluster = makeRootCluster();
+  const tshClientMock = await mockTshClient(tshDir, { clusters: [] });
+  const clusterStoreMock = mockClusterStore({ clusters: [cluster] });
+  const watcher = watchProfiles({
+    tshDirectory: tshDir,
+    tshClient: tshClientMock,
+    clusterStore: clusterStoreMock,
+    signal: abortController.signal,
+    debounceMs: testDebounceMs,
+  });
+
+  // Start the watcher by pulling the first value.
+  const firstEvent = watcher.next();
+  await fs.rm(tshDir, { recursive: true });
+  expect((await firstEvent).value).toEqual([{ op: 'removed', cluster }]);
+  // Clean up the store, so that we can detect a change.
+  clusterStoreMock.clearAll();
+
+  await fs.mkdir(tshDir);
+  await tshClientMock.insertOrUpdateCluster(cluster);
+  // The second event needs to wait for the dir to be detected.
+  jest.useFakeTimers();
+  const secondEvent = watcher.next();
+  // Polling uses 1 second interval.
+  jest.advanceTimersByTime(1000);
+  jest.useRealTimers();
+  expect((await secondEvent).value).toEqual([{ op: 'added', cluster }]);
+});
+
+test('max file system events count is restricted', async () => {
```

Comment:
This seems to work only on macOS but has some issues in our CI runner and on Windows.